backend/src/main/java/ca/bc/gov/hlth/hnweb/security/SecurityUtil.java

```diff
@@ -26,80 +27,131 @@ public class SecurityUtil {
 	private static final String CLAIM_RESOURCE_ACCESS = "resource_access";
 	private static final String CLAIM_SESSION_STATE = "session_state";
 	public static final String CLAIM_USERNAME = "preferred_username";
-	private static final String CLAIM_SUB = "sub";	//the Subject claim identifies the principal that is the subject of the JWT
+	private static final String CLAIM_SUB = "sub"; // the Subject claim identifies the principal that is the subject of
+													// the JWT
 	public static final String CLAIM_ORGANIZATION = "org_details";
-	
+
 	private static final String ORGANIZATION_ID = "id";
-	
+
 	private static final String USER_ROLES = "roles";
 
-    private static String KEYCLOAK_CLIENT;
-    
-    @Value("${spring.security.oauth2.resourceserver.jwt.audience}")
-    private void setKeycloakClientStatic(String keycloakClient){
-        SecurityUtil.KEYCLOAK_CLIENT = keycloakClient;
-    }
+	private static String KEYCLOAK_CLIENT;
+
+	private static SecurityProperties securityProperties;
+
+	@Autowired
+	public void setSecurityProperties(SecurityProperties properties) {
+		SecurityUtil.securityProperties = properties;
+	}
+
+	@Value("${spring.security.oauth2.resourceserver.jwt.audience}")
+	private void setKeycloakClientStatic(String keycloakClient) {
+		SecurityUtil.KEYCLOAK_CLIENT = keycloakClient;
+	}
 
-    public static UserInfo loadUserInfo() {
+	public static UserInfo loadUserInfo() {
 		Authentication auth = SecurityContextHolder.getContext().getAuthentication();
-		Jwt jwt = (Jwt)auth.getPrincipal();
-		
+		Jwt jwt = (Jwt) auth.getPrincipal();
+
 		UserInfo userInfo = new UserInfo();
 		userInfo.setOrganization(extractOrganization(jwt));
-		
+
 		List<String> roles = loadRoles(jwt);
-		userInfo.setRole(StringUtils.join(roles, " "));
+		userInfo.setRoles(roles);
 
 		userInfo.setSessionState(jwt.getClaim(CLAIM_SESSION_STATE));
 		userInfo.setUsername(jwt.getClaim(CLAIM_USERNAME));
 		userInfo.setUserId(jwt.getClaim(CLAIM_SUB));
-		
+
 		return userInfo;
 	}
-	
+
+	public static String loadSPGBasedOnTask(UserInfo userInfo, TransactionType transactionType) {
+		Map<String, List<String>> rolePermissions = securityProperties.getRolePermissions();
+		List<String> roles = userInfo.getRoles();
+		if (roles.size() == 1) {
+			return roles.get(0);
+		}
+		for (String role : roles) {
+			List<String> permissions = rolePermissions.get(role.toLowerCase());
+			switch (transactionType) {
+			case ENROLL_SUBSCRIBER:
+				if (permissions.contains("AddPermitHolderWOPHN") || permissions.contains("AddPermitHolderWithPHN")) {
+					return role;
+				}
+				break;
+			case GET_PERSON_DETAILS:
+				if (permissions.contains("AddPermitHolderWithPHN")) {
+					return role;
+				}
+				break;
+			case NAME_SEARCH:
+				if (permissions.contains("AddPermitHolderWOPHN")) {
+					return role;
+				}
+				break;
+			case CONTRACT_INQUIRY:
+				if (permissions.contains("ContractInquiry") || permissions.contains("GetGroupMembersContractAddress")) {
```

Comment:
GetGroupMembersContractAddress is not a valid permission. I know it's in SecurityConfig.java but there is an outstanding bug to fix that. Please use GetContractAddress instead.